hera_qm/xrfi.py

```diff
@@ -222,6 +225,266 @@ def robust_divide(num, den):
+    wgts = {} if wgts is None else deepcopy(wgts)
+    rfi_wgts = {}
+
+    # Infer int_count from data's metadata if not provided
+    if int_count is None:
+        int_time = 24 * 3600 * np.median(np.diff(data.times))
+        chan_res = np.median(np.diff(data.freqs))
+        int_count = int(int_time * chan_res)
+
+    array_weights = []
+
+    # Iterate through data
+    for k, v in data.items():
+        ant1, ant2 = split_bl(k)
+        # Only flag auto-correlations
+        if ant1 == ant2:
+            # If antenna is labeled "bad", skip and set weights to zeros
+            if antenna_class is not None and antenna_class.is_bad(ant1):
+                rfi_wgts[ant1] = np.zeros_like(v)
+                continue
+
+            # Estimate the noise standard deviation from data
+            noise = np.abs(v) / np.sqrt(int_count / 2)
+
+            # Run flagging algorithm
+            rfi_wgts[ant1] = alg_func(v, noise=noise, nsig=nsig, wgts=wgts.get(ant1, np.ones_like(v)), **flag_kwargs)
+            array_weights.append((1 - rfi_wgts[ant1]))
+
+    # Combine into array-wide weights
+    array_weights = np.mean(array_weights, axis=0)
+    array_weights = np.where(array_weights < array_wgts_threshold, 1, 0)
+    return rfi_wgts, array_weights
+
+def auto_rfi_checker(auto_wgts, array_wgts, good=(0, 3), suspect=(3, 6)):
+    """
+    Classifies ant-pols as good, suspect, or bad based on the fraction of channels flagged compared to the array average.
+    Ant-pols not in the good or suspect ranges will be labeled "bad". "auto_wgts" and "array_wgts" input parameters for this
+    function can be obtained from the "flag_autos" function.
+
+    Parameters:
+    ----------
+    auto_wgts: dictionary or DataContainer
+        Dictionary or DataContainer containing weights for each antenna with values of either zero or one with
+        zero identifying channels found to have RFI
+    array_weights: np.ndarray
+        Contains weights for the entire array averaged over all antennas
+    good: 2-tuple or list of 2-tuple bounds for ranges considered good. In units of relative
+            slope over the band.
+    suspect: 2-tuple or list of 2-tuple bounds for ranges considered suspect. Ant-pols
+        in both the good and suspect ranges will be labeled good.
+
+    Returns:
+        AntennaClassification with "good", "suspect", and "bad" ant-pols based on relative RFI flagging
+    """
+    # Get over-flagged fraction per-antenna where weights are not all zeros
+    flagged_fraction = np.array(
+        [
+            1 - np.sum(v * array_wgts) / np.sum(array_wgts)
+            for k, v in auto_wgts.items() if np.any(v)
+        ]
+    )
+
+    # Compute median and deviation
+    median = np.median(flagged_fraction)
+    std = np.median(np.abs(flagged_fraction - median)) / 0.6745
+
+    # Use standard deviation if std=0
+    if np.isclose(std, 0):
+        std = np.std(flagged_fraction)
+    
+    # For all antennas, determine a z-score
```

Comment:
One of the ideas behind ant_class is that we're trying to avoid relative cuts and only doing absolute cuts. 

Instead of z-scores, I think we should either cut on A) absolute fraction of the band flagged or B) absolute fraction of the band in excess of the array-averaged flagging. Perhaps you want to play around with a file and figure out which is better for catching outliers and not flagging good stuff?

Reply:
I ended up changing this to absolute fraction of the band flagged. It seemed to work fairly well at identifying outlier antennas.